### ISSUE TITLE
Optional properties fail types that do not have those properties

### ISSUE THREAD

**Issue by beaucollins** [facebook/flow#7707]
Flow version: 0.98.1

## Expected behavior

Width subtyping should allow types that _do not_ have optional properties of other type to be used in place of other type.

## Actual behavior

Types that do not have optional properties fail type checking even though the property is optional.

I expect this to be valid because `Thing`'s `notRequired` is optional so `OtherThing` shouldn't need it.

```js
type Thing = {
  notRequired?: string
}

type OtherThing = {
   name: string, 
}
const otherThing: OtherThing = {
}

function doit(thing: Thing) {
}

doit(otherThing);
```
* Link to Try-Flow or Github repo: https://flow.org/try/#0C4TwDgpgBAKgFgSwHYHMoF4oG8BQUpID2wAShAI4CuCAThACYD8AXFAM7A3Io4C+OOUJCgB5YHAg143DNjz4kAQwC2EVhy6oANFD44AxoSQcoxCVMSpWY89NSzc-AQDNKSfcARGo9QgmAAFOLcrHYoAJRy+Ho4vv4BZpJh4QDcQA

**Comment by dsainati1** [comment]
This is not a bug. You can think about option properties has having type `T | void`. That is, `{ x? : T}` is roughly equivalent to `{ x : T | void}`. If we allowed you to use width subtyping in the way you suggest, that would lead to the following unsound behavior:

https://flow.org/try/#0C4TwDgpgBAKgFgSwHYHMoF4oG8BQUpID2wAShAI4CuCAThACYD8AXFAM7A3Io4C+OOUJCgB5YHAg143DNjz4kAQwC2EVhy6oANFD4CAxoSQcoxCVMSpWY89NSysBYmSq0GUVgBYdS1R6gARIps9ABmAVD8AqGUSPrACEZQ9IQIwAAU4tysdigAlHL4ejgpaelmkrl5ANxAA

**Comment by beaucollins** [comment]
> If we allowed you to use width subtyping in the way you suggest, that would lead to the following unsound behavior:

How is this unsound?

**Comment by dsainati1** [comment]
Within the body of `doIt`, the type system believes that the `notRequired` property of `thing` is optionally a string. However, when we created `otherThing`, we set it to a number. If we were able to pass `otherThing` in to `doIt`, we would be in a situation where the type system believes `notRequired` is a string, but it's actually a number. If the implementation of `doIt` used a function that only worked on strings, there would be a runtime error. See https://flow.org/try/#0C4TwDgpgBAKgFgSwHYHMoF4oG8BQUpID2wAShAI4CuCAThACYD8AXFAM7A3Io4C+OOUJCgB5YHAg143DNjz4kAQwC2EVhy6oANFD4CAxoSQcoxCVMSpWY89NSysBYmSq0GUVgBYdS1R6gARIps9ABmAVD8AqGUSPrACEZQ9IQIwAAU4tysdigAlHIAkAihmZYoAHREpBTUdPR5uPj4hVmoVc61bvSybZXVLnUMFcCEADKEAO6SAMLBEOl5ANzy+PxRKWnpZpK5y0A for an example
